Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs

```diff
@@ -53,6 +53,100 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
         /// <inheritdoc />
         public virtual IMixedRealityController[] GetActiveControllers() => System.Array.Empty<IMixedRealityController>();
 
+        private struct PointerConfig
+        {
+            public PointerOption profile;
+            public Stack<IMixedRealityPointer> cache;
+        }
+
+        private PointerConfig[] pointerConfigurations = System.Array.Empty<PointerConfig>();
+
+        private class PointerEqualityComparer : IEqualityComparer<IMixedRealityPointer>
+        {
+            private static PointerEqualityComparer defaultComparer;
+
+            internal static PointerEqualityComparer Default
+            {
+                get =>  defaultComparer ?? (defaultComparer = new PointerEqualityComparer());
+            }
+
+            /// <summary>
+            /// Check that references equals for two pointers
+            /// </summary>
+            public bool Equals(IMixedRealityPointer p1, IMixedRealityPointer p2)
+            {
+                return ReferenceEquals(p1, p2);
+            }
+
+            /// <summary>
+            /// Unity objects have unique equals comparison and to check keys in a dictionary, 
+            /// we want the hascode match to be Unity's unique InstanceID to compare objects
+            /// </summary>
+            public int GetHashCode(IMixedRealityPointer pointer)
+            {
+                if (pointer is MonoBehaviour pointerObj)
+                {
+                    return pointerObj.GetInstanceID();
+                }
+                else
+                {
+                    return pointer.GetHashCode();
+                }
+            }
+        }
+
+        // Active pointers associated with the config index they were spawned from
+        private Dictionary<IMixedRealityPointer, uint> activePointersToConfig 
+            = new Dictionary<IMixedRealityPointer, uint>(PointerEqualityComparer.Default);
+
```

Comment:
Another point to consider when doing this sort of change - I assume that we're gonna have bugs (because there are always bugs, it's a fact of life!)

It would be interesting to have a boolean/flag/control mechanism (maybe on this class, maybe on the profile itself) that will either turn on or turn off this behavior. This way we can really easily test if/when there are issues, if they are caused by recycling of pointers. i.e. if someone notices a bug, they can try turning off this new feature and see if the issue still repros, and if not, we have a better idea of what to look at.

Obviously such a flag would be removed soonish after in a future release, but in the interim these control mechanisms are really helpful for issue/feature isolation.